rclcpp_action/include/rclcpp_action/client.hpp

```diff
@@ -632,22 +625,28 @@ class Client : public ClientBase
     using GoalResultRequest = typename ActionT::Impl::GetResultService::Request;
     auto goal_result_request = std::make_shared<GoalResultRequest>();
     goal_result_request->goal_id.uuid = goal_handle->get_goal_id();
-    this->send_result_request(
-      std::static_pointer_cast<void>(goal_result_request),
-      [goal_handle, this](std::shared_ptr<void> response) mutable
-      {
-        // Wrap the response in a struct with the fields a user cares about
-        WrappedResult wrapped_result;
-        using GoalResultResponse = typename ActionT::Impl::GetResultService::Response;
-        auto result_response = std::static_pointer_cast<GoalResultResponse>(response);
-        wrapped_result.result = std::make_shared<typename ActionT::Result>();
-        *wrapped_result.result = result_response->result;
-        wrapped_result.goal_id = goal_handle->get_goal_id();
-        wrapped_result.code = static_cast<ResultCode>(result_response->status);
-        goal_handle->set_result(wrapped_result);
-        std::lock_guard<std::mutex> lock(goal_handles_mutex_);
-        goal_handles_.erase(goal_handle->get_goal_id());
-      });
+    try {
+      this->send_result_request(
+        std::static_pointer_cast<void>(goal_result_request),
+        [goal_handle, this](std::shared_ptr<void> response) mutable
+        {
+          // Wrap the response in a struct with the fields a user cares about
+          WrappedResult wrapped_result;
+          using GoalResultResponse = typename ActionT::Impl::GetResultService::Response;
+          auto result_response = std::static_pointer_cast<GoalResultResponse>(response);
+          wrapped_result.result = std::make_shared<typename ActionT::Result>();
+          *wrapped_result.result = result_response->result;
+          wrapped_result.goal_id = goal_handle->get_goal_id();
+          wrapped_result.code = static_cast<ResultCode>(result_response->status);
+          goal_handle->set_result(wrapped_result);
+          std::lock_guard<std::mutex> lock(goal_handles_mutex_);
+          goal_handles_.erase(goal_handle->get_goal_id());
+        });
+    } catch (rclcpp::exceptions::RCLError &) {
+      goal_handle->invalidate();
```

Comment:
@jacobperron hmm, my only concern here is that the caller won't know that the result request went wrong till much later. It also means a caller can't cancel the goal just sent.

---

What if we just eat up the exception? It's not ideal, but `ClientGoalHandle::is_result_aware()` can still return `false` and `ClientGoalHandle::async_get_result()` (which will eventually be called anyways) can throw.

Reply:
Perhaps it's not as ergonomic, but the caller can still immediately check the future returned by `Client::async_get_result`, and any error that occurred when sending the result request should be visible.

>  It also means a caller can't cancel the goal just sent.

This is a valid concern. I think what you're proposing is to remove the erasure logic below so that the user can still interact with the goal handle (even though it failed to get the result). I think calling `invalidate()` still makes sense so that calling `Client::async_get_result` still produces an exception.

Reply:
The other option I hadn't considered was to not handle the exception at all. This way we leave it up to the caller how to handle the exception (and they know something went wrong right away).

Reply:
Right. If `ClientGoalHandle::make_result_aware` fails, the goal handle will not be result aware. Any attempt to get the future [will result in an exception](https://github.com/ros2/rclcpp/blob/0313417f0213f1741465978c91ff18ba4f99ab55/rclcpp_action/include/rclcpp_action/client_goal_handle_impl.hpp#L63-L65) that the user can handle.

Reply:
Take a look at 81239f1

As it is currently implemented, the user will get the [exception set in the future](https://github.com/ros2/rclcpp/blob/0313417f0213f1741465978c91ff18ba4f99ab55/rclcpp_action/include/rclcpp_action/client_goal_handle_impl.hpp#L145) (because we set result awareness to `true` above, before sending a result request).

But, I think it makes sense to throw the error as earlier if possible, so I've make a small change make this happen: d444e09

Reply:
Do we have to do anything besides making the goal handle not result aware? Would it be wrong if the user attempts to request the result again?

Reply:
I don't think there's anything else to do :thinking: If a user attempts to request the result I'm not actually sure what would happen, but I guess it shouldn't be allowed considering the result future would have already been set to an exception. In retrospect, I think it keeps the logic simpler to let the exception live in the future (1176dad).
Currently, the user is not able to request the result more than once (even if it is successful).

Reply:
Hmm, I see. We're not dealing with exceptions correctly e.g. depending on where `Client::async_get_result()` throws, the handle may be tagged as result aware (banning future result requests) or not. 

In my mind, a user that explicitly makes a goal handle result aware should get an exception if something goes wrong, and the goal handle should remain result unaware. That is, `Client::async_get_result()` should offer strong exception guarantees. It shouldn't be different for a user that implicitly makes a goal handle result aware by passing in a result callback to `Client::async_send_goal()`. If we don't make the latter throw (and there're valid arguments against), by keeping the goal handle result unaware we allow the user to check `ClientGoalHandle::is_result_aware()` or handle the exception that `ClientGoalHandle::async_get_result()` will throw. Does that make sense?

To invalidate the goal handle improves the situation quite a bit, but it delays error generation (until `future::get`, when it could show itself before spinning), it induces an spurious status change (to UNKNOWN until another status message comes along), and it prevents any retries on an otherwise unaware goal handle. 

But I won't block you. CI's green and this is a fairly unusual case :)

Reply:
Yeah, that makes sense to me. I've made changes so that in the event that a result request fails, `ClientGoalHandle::is_result_aware()` should return `false` and `Client::async_get_result()` raises an exception (9cb1776).

I couldn't find a nice way to do it with the existing state information, so I've introduced a new member to `ClientGoalHandle` for checking if `invalidate()` was called, which doubles as an improved error message.

PTAL

Reply:
Neat! Two questions. Could a goal handle be invalidated twice? Should we do something about the spurious UNKNOWN status on invalidation?

Reply:
Good questions. I think we certainly want to guard against multiple calls invalidate (see b145c73). As an aside, I wish it was easier to test this stuff :upside_down_face: It might be worth revisiting the architecture to make validation easier.

I'm not sure what to do about the UNKNOWN status. We could remove setting it when the goal is invalidated. Really I think the status should be reporting the latest state from the action server. Since this is keeping the same behavior (ie. when the client is destroyed), I'll defer deciding to remove the unknown status to another PR.